### PR TITLE
WIP: Disloc kink structure generation

### PR DESCRIPTION
**Pull request by thomas-rocke**
Tools for building kink structures both in cylinder configs, and in dislocation quadrupoles.

Planned Features:
- [x] Compression of disloc.unit_cell to use minimum number of atoms in cubic cell
- [x] Small utility for calculating coordination number
- [x] Dislocation Cylinders with arbitrary kink networks in x direction
- [x] Kink Quadrupoles for single kink in each direction
- [x] Displacement interpolation for quadrupoles
- [ ] Docs covering glide & kink in more detail
- [ ] Unittests covering kink generation

Possible extra features:
- [x] Multi-kink quadrupoles (i.e. kink out by 2 glides instantaneously)
- [x] Displacement interpolation for kink cylinders
- [x] Kink networks in quadrupoles?
- [ ] More useful coordination function
    - [ ] Multiple radii at once, not just single radius
- [ ] Increase efficiency of quad kink slicing algorithm
- [ ] Networks of kinks in multiple glide planes for Screw dislocations
- [ ] Rectangular periodic array of (non-quadrupolar) dislocations

### CODE REVIEW

File: matscipy/neighbours.py [hunk 1]
```diff
@@ -940,3 +940,18 @@ def find_common_neighbours(i_n, j_n, nat):
...
+
+
+def coordination(structure, cutoff):
+    '''
+    Get coordination number of each atom in structure
+    based on the number of neighbours within cutoff
+
+    structure: ase Atoms object
+        Structure to take coordination of
+    cutoff: float
```
Comment (jameskermode) [inline matscipy/neighbours.py:952]: doc fix: could be anything that neigbour_list accepts, e.g. a dict

File: matscipy/utils.py [hunk 2]
```diff
@@ -1,12 +1,14 @@
 import warnings
 import numpy as np
+from ase.utils.structure_comparator import SymmetryEquivalenceCheck
```
Comment (jameskermode) [inline matscipy/utils.py:3]: not being used

Comment (jameskermode) [reply matscipy/utils.py:3]: (check this)

File: matscipy/utils.py [hunk 3]
```diff
@@ -112,6 +119,92 @@ def validate_cubic_cell(a, symbol="w", axes=None, crystalstructure=None, pbc=Tru
...
+    '''
+    Search for a condensed representation of atoms
+    Equivalent to attempting to undo a supercell
+
+    atoms: ase Atoms object
+        structure to condense
+    precision: int
+        Number of decimal places to use in determining whether scaled positions are equal
+
+    returns a condensed copy of atoms, if such a condensed representation is found. Else returns a copy of atoms
```
Comment (jameskermode) [inline matscipy/utils.py:133]: maybe return a bool for whether it was modified or not and new structure?

File: matscipy/utils.py [hunk 4]
```diff
@@ -112,6 +119,92 @@ def validate_cubic_cell(a, symbol="w", axes=None, crystalstructure=None, pbc=Tru
 
     return alat, unit_cell
 
+
+def find_condensed_repr(atoms, precision=2):
+    '''
```
Comment (jameskermode) [inline matscipy/utils.py:124]: pep8: I think triple quoted stings are supposed to be """ terminated

File: matscipy/utils.py [hunk 5]
```diff
@@ -112,6 +119,92 @@ def validate_cubic_cell(a, symbol="w", axes=None, crystalstructure=None, pbc=Tru
...
+    ]
+
+    dirs = cart_directions[axis]
+
+    ats = atoms.copy()
+    
+    # Find all atoms which are in line with the 0th atom
+    p = np.round(ats.get_scaled_positions(), precision)
+
+    p_diff = (p - p[0, :]) % 1
```
Comment (jameskermode) [inline matscipy/utils.py:172]: maybe % 1.0 better if it's being treated as a float?

File: matscipy/dislocation.py [hunk 6]
```diff
@@ -2607,6 +2609,8 @@ def _build_bulk_cyl(self, radius, core_positions, fix_rad, extension,
             Will add a fictitious core at position 
             core_positions[i, :] + extension[i, :], for the 
             purposes of adding extra atoms
+        cyl_mask: array of bool
+            Optional ovverride for atomic mask to convert supercell into cyl
```
Comment (jameskermode) [inline matscipy/dislocation.py:2613]: expand docs on reference frame

File: matscipy/dislocation.py [hunk 7]
```diff
@@ -2616,35 +2620,53 @@ def _build_bulk_cyl(self, radius, core_positions, fix_rad, extension,
...
+        # Validate core position, extension, and fixed_points args
+        errs = []
+        for argname, arg in [["core_position", core_positions], ["extension", extension], ["fixed_points", fixed_points]]:
+            if type(arg) == np.ndarray:
+                if arg.shape[-1] != 3:
+                    # Needs to be 3d vectors
+                    errs.append(f"Argument {argname} misspecified. Should be an array of shape (N, 3)")
+                if len(arg.shape) == 1:
+                    # Convert all arrays to 2D
+                    arg.reshape((np.newaxis, arg.shape[-1]))
```
Comment (jameskermode) [inline matscipy/dislocation.py:2635]: consider `np.atleast_2d`?

File: matscipy/dislocation.py [hunk 8]
```diff
@@ -3029,6 +3089,77 @@ def build_impurity_cylinder(self, disloc, impurity, radius,
...
+        # ReLU interpolation between kink1 & kink2 displacements 
+        smooth_facs = (x > h - sw) * (x + sw - h) / sw
+
+        smoothed_disp = dr1 + smooth_facs[:, np.newaxis] * dr_forward
+
+        kink.positions += smoothed_disp
+
+        return kink
+    
+    def build_kink_cyl(self, kink_map=[0, 1], smooth_width=None, *args, **kwargs):
```
Comment (jameskermode) [inline matscipy/dislocation.py:3123]: don't use mutable objects as default arguments

File: matscipy/dislocation.py [hunk 9]
```diff
@@ -3416,7 +3576,15 @@ def build_cylinder(self, radius, partial_distance=0,
...
-        return bulk, disloc
+        
+        out = [bulk, disloc]
+
+        if return_cyl_mask:
+            out.append(cyl_mask)
+        if return_fix_mask:
+            out.append(fix_mask)
+
+        return out
```
Comment (jameskermode) [inline matscipy/dislocation.py:3587]: convert to tuple before returning

File: matscipy/dislocation.py [hunk 10]
```diff
@@ -3801,7 +3969,86 @@ def build_glide_quadrupoles(self, nims, invert_direction=False, glide_left=True,
...
+
+        bulk_coord = [np.min(coordination(bulk, cutoff * self.alat)) for cutoff in cutoffs]
+
+        full_mask = np.ones((len(tilt_bulk)), dtype=bool)
+        atom_heights = np.round(p, precision)[:, 2]
+
+        for i in range(int(cell[2, 2] / step_size)):
+            # Check if tilt_bulk has same 1st, 2nd, & 3rd neighbour coordination
+            # as normal bulk
+            # TODO: This currently does not work for BCCEdge111barDislocation
```
Comment (jameskermode) [inline matscipy/dislocation.py:4031]: can you raise an exception?

**Comment by pgrigorev** [comment]
Hi @thomas-rocke, really fantastic work. Tutorial on dislocation glide was for long time on my to do list, so I am very happy you did it. Visualisation of kink looks amazing. Also kink glide methods look very useful, can help me with few annoying things. Well done!

**Pull request merged**